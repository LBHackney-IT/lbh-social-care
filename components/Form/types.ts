/* eslint-disable @typescript-eslint/no-explicit-any */

import { InputHTMLAttributes } from 'react';
import { RegisterOptions } from 'react-hook-form';
import TextInput from './TextInput/TextInput';

type LabelSize = 's' | 'm' | 'l' | 'xl';
type Width = 30 | 20 | 10 | 5 | 4 | 3 | 2;
type GovGrid =
  | 'full'
  | 'half'
  | 'one-half'
  | 'one-third'
  | 'two-thirds'
  | 'one-quarter'
  | 'three-quarters';

export type Option = string | { text: string; value: string };

type FuncOption = (arg0: Record<string, string>) => Option[] | null;

export type Options = FuncOption | Option[];

export interface GenericField {
  name: string;
  label?: string;
  rules?: RegisterOptions;
  register?: (arg0: RegisterOptions) => any;
  error?: { message?: string };
  required?: boolean | string;
  hint?: string;
  labelSize?: LabelSize;
  conditionalRender?: (arg0: Record<string, any>) => boolean;
  isMulti?: boolean;
  isMultiInit?: boolean;
  isMultiTrigger?: string;
}

export interface Checkbox extends GenericField {
  options?: Option[];
}

interface StepCheckbox extends Omit<Checkbox, 'options'> {
  component: 'Checkbox';
  options?: Options;
}

type TextInputType = GenericField & InputHTMLAttributes<HTMLInputElement>;

export interface TextInput extends TextInputType {
  inputClassName?: string;
  width?: Width;
}

interface StepTextInput extends TextInput {
  component: 'TextInput';
}

export type TextInputNoType = Omit<TextInput, 'type'>;

interface StepTextInputNoType extends TextInputNoType {
  component: 'NumberInput' | 'DatePicker' | 'PhoneInput' | 'EmailInput';
}

type TextAreaType = GenericField & InputHTMLAttributes<HTMLTextAreaElement>;

export interface TextArea extends TextAreaType {
  width?: Width;
  rows?: number;
}

interface StepTextArea extends TextArea {
  component: 'TextArea';
}

type SelectType = GenericField &
  Omit<InputHTMLAttributes<HTMLSelectElement>, 'onChange'>;

export interface Select extends SelectType {
  options: Option[];
  placeHolder?: string;
  ignoreValue?: boolean;
  isUnselectable?: boolean;
  width?: Width;
  govGrid?: GovGrid;
  onChange?: (arg0: string) => string | void;
}

interface StepSelect extends Omit<Select, 'options'> {
  component: 'Select';
  options: Options;
}

type RadiosType = GenericField & InputHTMLAttributes<HTMLInputElement>;

export interface Radios extends RadiosType {
  options?: Option[];
  isRadiosInline?: boolean;
  children?: React.ReactNode;
}

interface StepRadios extends Omit<Radios, 'options'> {
  component: 'Radios';
  options?: Options;
}

export interface ObjectInput extends GenericField {
  isMulti?: boolean;
  isInline?: boolean;
  summaryInline?: boolean;
  components: Array<formBasicComponentStep>;
}

interface StepObjectInput extends ObjectInput {
  component: 'ObjectInput';
}

export interface AddressLookup extends GenericField {
  supportManualEntry?: boolean;
}

interface StepAddressLookup extends AddressLookup {
  component: 'AddressLookup';
}

export interface DateInput extends GenericField {
  format?: 'US' | 'EU';
}

interface StepDateInput extends DateInput {
  component: 'DateInput';
}

type formBasicComponentStep =
  | StepTextInput
  | StepTextInputNoType
  | StepTextArea
  | StepCheckbox
  | StepSelect
  | StepRadios
  | StepAddressLookup
  | StepDateInput;

export type FormComponentStep =
  | formBasicComponentStep
  | StepObjectInput
  | JSX.Element;

export interface FormStep {
  id: string;
  title: string;
  components: Array<FormComponentStep>;
  isMulti?: boolean;
  conditionalRender?: (arg0: Record<string, unknown>) => boolean;
}

export interface FormConfig {
  title: string;
  path: string;
  successMessage?: string;
  steps: Array<FormStep>;
}
