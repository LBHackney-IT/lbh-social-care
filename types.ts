import { AxiosError } from 'axios';

export type ErrorAPI = AxiosError;

export interface Address {
  address: string;
  postcode: string;
  uprn: string;
}

export type AgeContext = 'A' | 'B' | 'C' | undefined;

export interface Allocation {
  id: number;
  caseStatus: 'Closed' | 'Open';
  allocatedWorkerTeam: string;
  allocatedWorker: string;
  allocationStartDate: string;
  allocationEndDate?: string;
  workerType: string;
  personId: number;
  personName: string;
  personDateOfBirth: string;
  personAddress: string;
}

export interface AllocationData {
  allocations: Allocation[];
}

interface CaseFormDataBase {
  mosaic_id: number;
  first_name: string;
  last_name: string;
  worker_email: string;
  form_name_overall: string;
  form_name: string;
  context_flag: AgeContext;
  date_of_event: string | Date;
  timestamp: string | Date;
  is_historical?: true | false;
  note?: string;
  date_of_birth?: string;
  form_url?: string;
  case_note_title?: string;
  case_note_description?: string;
}

export interface AllocationCaseFormData extends CaseFormDataBase {
  form_name_overall: 'API_Allocation';
  allocation_id: number;
  created_by: string;
}

export interface DeallocationCaseFormData extends CaseFormDataBase {
  form_name_overall: 'API_Deallocation';
  allocation_id: number;
  deallocation_reason: string;
  created_by: string;
}

export interface WarningNoteCaseFormData extends CaseFormDataBase {
  form_name_overall: 'API_WarningNote';
  warning_note_id: number;
}

export interface HistoricCaseData {
  title: string;
  formName?: string;
  content?: string;
  officerName?: string;
  officerEmail?: string;
  dateOfEvent: string;
}

export interface HistoricVisitData {
  visitId: number;
  personId: number;
  visitType: string;
  plannedDateTime: string;
  actualDateTime: string;
  createdByName: string;
  createdByEmail: string;
  reasonNotPlanned?: string;
  reasonVisitNotMade?: string;
  seenAloneFlag: boolean;
  completedFlag: boolean;
}

export type CaseFormData =
  | CaseFormDataBase
  | AllocationCaseFormData
  | DeallocationCaseFormData
  | WarningNoteCaseFormData;

export interface Case {
  recordId: string;
  formName: string;
  personId: number;
  firstName?: string;
  lastName?: string;
  dateOfBirth?: string;
  officerEmail?: string;
  caseFormUrl?: string;
  caseFormTimestamp?: string;
  dateOfEvent?: string;
  caseFormData: CaseFormData;
}

export interface CaseData {
  cases: Case[] | [];
  nextCursor?: number;
}

export interface LegacyResident {
  mosaicId: number;
  firstName: string;
  lastName: string;
  uprn?: string;
  dateOfBirth: string;
  ageContext: AgeContext;
  gender: string;
  nationality?: string;
  nhsNumber: string;
  restricted?: 'Y' | 'N';
  address?: {
    address: string | null;
    postcode: string | null;
    uprn?: string;
  };
  phoneNumber?: Array<{
    phoneNumber: string;
    phoneType: string;
  }>;
}

export interface ResidentsAPI {
  residents: LegacyResident[] | [];
  nextCursor?: string;
}

/**
 * This should replace LegacyResident
 * When /residents is going to be aligned with /residents/:id
 */
export interface Resident {
  id: number;
  firstName: string;
  lastName: string;
  gender: string;
  contextFlag: AgeContext;
  createdBy: string;
  otherNames: Array<{
    firstName: string;
    lastName: string;
  }>;
  phoneNumbers: Array<{
    number: string;
    type: string;
  }>;
  title?: string;
  dateOfBirth?: string;
  ethnicity?: string;
  firstLanguage?: string;
  religion?: string;
  sexualOrientation?: string;
  nhsNumber?: number;
  emailAddress?: string;
  preferredMethodOfContact?: string;
  restricted?: 'Y' | 'N';
  dateOfDeath?: string;
  address?: {
    address: string;
    postcode: string;
    uprn?: string;
  };
}

export interface User {
  name: string;
  email: string;
  permissionFlag: AgeContext;
  hasDevPermissions: boolean;
  hasAdminPermissions: boolean;
  hasAdultPermissions: boolean;
  hasChildrenPermissions: boolean;
  hasAllocationsPermissions?: boolean;
  hasUnrestrictedPermissions?: boolean;
  isAuthorised: boolean;
}

export interface Team {
  id: number;
  name: string;
}

export interface Worker {
  id: number;
  email: string;
  firstName: string;
  lastName: string;
  role: string;
  allocationCount: number;
  teams: Team[];
}

export interface RelationshipPerson {
  id: number;
  firstName: string;
  lastName: string;
}

export interface Relationship {
  parents: Array<RelationshipPerson>;
  siblings: Array<RelationshipPerson>;
  children: Array<RelationshipPerson>;
  other: Array<RelationshipPerson>;
}

export interface RelationshipData {
  personId: number;
  personalRelationships: Relationship;
}

interface BaseNote {
  id: number;
  noteType: string;
  createdBy: string;
  startDate: string;
  reviewDate: string;
  endDate?: string;
  endedBy?: string;
  reviewedDate?: string;
  reviewedBy?: string;
  nextReviewDate?: string;
  notes: string;
  managerName: string;
  discussedWithManagerDate: string;
  status: 'closed' | 'open';
  warningNoteReviews: Array<ReviewedNote>;
}

interface DisclosedNote extends BaseNote {
  disclosedWithIndividual: true;
  disclosedDetails: string;
  disclosedDate: string;
  disclosedHow: 'Verbal' | 'Written' | 'Verbal / Written';
}

interface UndisclosedNote extends BaseNote {
  disclosedWithIndividual: false;
  undisclosedDetails: string;
}

interface ReviewedNote {
  id: number;
  warningNoteId: number;
  reviewDate: string;
  disclosedWithIndividual: boolean;
  reviewNotes: string;
  managerName: string;
  discussedWithManagerDate: string;
  createdAt: string;
  createdBy: string;
  lastModifiedAt?: string;
  lastModifiedBy: string;
}

export type WarningNote = DisclosedNote | UndisclosedNote;

export interface WarningNotes {
  notes: Array<WarningNote>;
}
