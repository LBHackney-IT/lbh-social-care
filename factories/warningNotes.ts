import { Factory } from 'fishery';

import { WarningNote } from 'types';

export const warningNoteFactory = Factory.define<WarningNote>(
  ({ sequence }) => ({
    id: sequence,
    type: 'Risk to Staff',
    startDate: new Date(2020, 11, 22),
    createdBy: 'Foo',
    disclosedWithIndividual: false,
    undisclosedDetails: 'nope',
    discussedWithManager: 'the big boss',
    discussedWithManagerDate: new Date(2020, 11, 12),
    reviewDate: new Date(2021, 1, 1),
    notes: 'a lot to talk about',
    status: 'open',
    reviews: [],
  })
);

export const mockedWarningNote = [
  warningNoteFactory.build({
    type: 'Risk to Adults',
    reviewedDate: new Date(2020, 11, 13),
    reviewedBy: 'Bar',
    disclosedDate: new Date(2020, 11, 12),
    disclosedHow: 'Written',
    disclosedWithIndividual: true,
    disclosedDetails: 'lorem ipsum',
    reviewDate: new Date(2020, 11, 31),
    notes: 'i am a note',
  }),
  warningNoteFactory.build(),
];
