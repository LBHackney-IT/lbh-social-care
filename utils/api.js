import { getQueryString } from 'utils/urls';

export const getInfiniteKey = (baseUrl, key, params) => (
  pageIndex,
  previousPageData
) => {
  if (previousPageData && !previousPageData[key]) return null;
  if (pageIndex === 0)
    return `${baseUrl}${params ? `?${getQueryString(params)}` : ''}`;
  return `${baseUrl}?${getQueryString({
    ...params,
    cursor: previousPageData.nextCursor,
  })}`;
};
