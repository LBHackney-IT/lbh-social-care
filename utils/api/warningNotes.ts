import axios from 'axios';
import useSWR, { SWRResponse } from 'swr';

import type { WarningNote, ErrorAPI } from 'types';

export const useWarningNotes = (
  id: number
): SWRResponse<WarningNote[], ErrorAPI> =>
  useSWR(`/api/residents/${id}/warningnotes`);

export const useWarningNote = (
  warningNoteId: number
): SWRResponse<WarningNote, ErrorAPI> =>
  useSWR(`/api/warningnotes/${warningNoteId}`);

export const addWarningNote = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.post(`/api/warningnotes`, formData);
  return data;
};

export const updateWarningNote = async (
  noteId: number,
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/warningnotes/${noteId}`, formData);
  return data;
};
