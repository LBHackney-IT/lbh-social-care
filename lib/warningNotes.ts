import axios from 'axios';

import { mockedWarningNote } from 'factories/warningNotes';
import { WarningNote } from 'types';

const { ENDPOINT_API, AWS_KEY } = process.env;

const headers = { 'x-api-key': AWS_KEY };

export const getWarningNotesByResident = async (
  personId: number
): Promise<WarningNote[] | []> => {
  const { data } = await axios.get(
    `${ENDPOINT_API}/residents/${personId}/warningnotes`,
    {
      headers,
    }
  );
  return data.warningNotes;
};

export const getWarningNote = async (
  warningNoteId: number
): Promise<WarningNote | undefined> => {
  return await new Promise((resolve) =>
    setTimeout(
      () => resolve(mockedWarningNote.find(({ id }) => id === warningNoteId)),
      1000
    )
  );
};

export const addWarningNote = async (
  params: Record<string, unknown>
): Promise<void> => {
  await axios.post(`${ENDPOINT_API}/warningnotes`, params, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
};

export const patchWarningNote = async (
  personId: number,
  params: Record<string, unknown>
): Promise<void> => {
  await axios.patch(`${ENDPOINT_API}/warningnotes/${personId}`, params, {
    headers: { ...headers, 'Content-Type': 'application/json' },
  });
};
