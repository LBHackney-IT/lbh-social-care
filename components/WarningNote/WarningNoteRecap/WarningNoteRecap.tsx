import Spinner from 'components/Spinner/Spinner';
import Summary from 'components/Summary/Summary';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import { useWarningNote } from 'utils/api/warningNotes';
import { formStepsAdult, formStepsChild } from 'data/forms/warning-note';
import { Resident } from 'types';
import reviewFormSteps from 'data/forms/warning-note-review';
import { FormComponentStep, FormStep } from 'components/Form/types';

export interface Props {
  person: Resident;
  warningNoteId: number;
  formName?: string;
}

export const modifyReviewSummary = (): FormStep[] => {
  const reviewedBy: FormComponentStep = {
    component: 'TextInput',
    name: 'lastModifiedBy',
    label: 'Review done by',
  };

  const endDate: FormComponentStep = {
    component: 'DateInput',
    name: 'endDate',
    label: 'End Date',
  };

  const nextReviewDateWithoutReviewDecision: FormComponentStep = {
    component: 'DateInput',
    name: 'nextReviewDate',
    label: 'Next Review Date',
  };

  const updatedReviewFormSteps = reviewFormSteps;

  updatedReviewFormSteps[0].components.pop();
  updatedReviewFormSteps[0].components.splice(1, 0, reviewedBy);
  updatedReviewFormSteps[0].components.push(
    endDate,
    nextReviewDateWithoutReviewDecision
  );

  return updatedReviewFormSteps;
};

const reviewedNoteSummary = modifyReviewSummary();

const WarningNoteRecap = ({
  person,
  warningNoteId,
  formName,
}: Props): React.ReactElement => {
  const { data: warningNote, error: warningNoteError } =
    useWarningNote(warningNoteId);

  if (warningNoteError) {
    return <ErrorMessage />;
  }
  if (!warningNote) {
    return <Spinner />;
  }

  const displayReviews =
    Array.isArray(warningNote.warningNoteReviews) &&
    warningNote.warningNoteReviews.length > 0 &&
    formName != 'Warning Note Created';

  return (
    <>
      {displayReviews &&
        warningNote.warningNoteReviews
          .slice(0)
          .reverse()
          .map((review) => {
            return (
              <Summary
                key={review.id}
                formData={{
                  ...review,
                  endDate: warningNote.endDate,
                  nextReviewDate: warningNote.nextReviewDate,
                  person,
                }}
                formSteps={reviewedNoteSummary.map((step) => {
                  return {
                    ...step,
                    title: `Warning Review Details ${new Date(
                      review.reviewDate
                    ).toLocaleDateString('en-GB')}`,
                  };
                })}
                formPath={`/people/:peopleId/warning-notes/:warningNoteId`}
              />
            );
          })}
      <Summary
        formData={{
          ...warningNote,
          disclosedWithIndividual: warningNote.disclosedWithIndividual
            ? 'Yes'
            : 'No',
        }}
        formSteps={person.contextFlag === 'A' ? formStepsAdult : formStepsChild}
        formPath={`/people/:peopleId/warning-notes/:warningNoteId`}
      />
    </>
  );
};

export default WarningNoteRecap;
