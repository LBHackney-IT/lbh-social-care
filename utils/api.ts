import { getQueryString } from 'utils/urls';

interface PageData {
  nextCursor?: string;
  [key: string]: unknown;
}

export const getInfiniteKey = (
  baseUrl: string,
  key: string,
  params: Record<string, unknown>
) => (pageIndex: number, previousPageData: PageData): string | null => {
  if (previousPageData && !previousPageData[key]) return null;
  if (pageIndex === 0)
    return `${baseUrl}${params ? `?${getQueryString(params)}` : ''}`;
  return `${baseUrl}?${getQueryString({
    ...params,
    cursor: previousPageData.nextCursor,
  })}`;
};
