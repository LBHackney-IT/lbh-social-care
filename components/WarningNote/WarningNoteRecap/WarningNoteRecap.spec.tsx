import { render } from '@testing-library/react';

import WarningNoteRecap from './WarningNoteRecap';
import * as warningNotes from 'utils/api/warningNotes';
import { mockedWarningNote } from 'fixtures/warningNotes.fixtures';
import { AgeContext } from 'types';

jest.mock('components/Spinner/Spinner', () => () => 'MockedSpinner');

describe(`WarningNoteRecap`, () => {
  jest.spyOn(warningNotes, 'useWarningNote').mockImplementation(() => ({
    data: mockedWarningNote.find(({ id }) => id === props.warningNoteId),
    mutate: jest.fn(),
    revalidate: jest.fn(),
    isValidating: false,
  }));
  const props = {
    person: {
      dateOfBirth: '2020-11-13',
      firstName: 'Ciasom',
      lastName: 'Tesselate',
      mosaicId: 44000000,
      nhsNumber: '12345',
      ageContext: 'A' as AgeContext,
      gender: 'F',
    },
    warningNoteId: 123,
  };

  it('should render correctly', async () => {
    const { asFragment } = render(<WarningNoteRecap {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
