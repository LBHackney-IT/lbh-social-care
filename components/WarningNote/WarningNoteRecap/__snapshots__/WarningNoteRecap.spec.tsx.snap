// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarningNoteRecap should render correctly 1`] = `
<DocumentFragment>
  <div
    class="govuk-!-margin-bottom-7"
  >
    <div
      class="lbh-table-header"
    >
      <h3
        class="lbh-heading-h2 section-heading"
      >
        Warning details
      </h3>
    </div>
    <dl
      class="govuk-summary-list  lbh-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Select warning type
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Risk to Adults
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Start date
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          22/12/2020
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Review / end date
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          31/12/2020
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Is the individual aware of the Warning Note and Review Date?
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Yes
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Date informed
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          12/12/2020
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          How informed
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          written
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Details of disclosure to individual
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <div>
            <div
              class="container"
              style="max-height: 95px;"
            >
              <div>
                <pre
                  class="formattedTextArea"
                >
                  lorem ipsum
                </pre>
              </div>
            </div>
          </div>
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Warning narrative and risks notes
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          <div>
            <div
              class="container"
              style="max-height: 95px;"
            >
              <div>
                <pre
                  class="formattedTextArea"
                >
                  i am a note
                </pre>
              </div>
            </div>
          </div>
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Managerâ€™s name
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          the big boss
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Date discussed with manager
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          12/12/2020
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;
